server/src/main/java/org/opensearch/common/remote/RemoteWritableEntityStore.java

```diff
@@ -0,0 +1,28 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.common.remote;
+
+import org.opensearch.core.action.ActionListener;
+
+import java.io.IOException;
+
+/**
+ * An interface to read/write an object from/to a remote storage. This interface is agnostic of the remote storage type.
+ *
+ * @param <T> The object type which can be uploaded to or downloaded from remote storage.
+ * @param <U> The wrapper entity which provides methods for serializing/deserializing entity T.
+ */
+public interface RemoteWritableEntityStore<T, U extends RemoteWriteableEntity<T>> {
+
+    public void writeAsync(U entity, ActionListener<Void> listener);
+
+    public U read(U entity) throws IOException;
```

Comment:
The interface doesn't look right to me. You are trying to read an entity `U` when you already have a partial `U`

Reply:
The method argument U contains the information to read T from remote. After reading it, it should set object T in U and return.

Reply:
I understand what is unclear is that what is a mandatory property for an entity then. From what I can grok, we are using blob path as a property to construct the entity and then set the actual response once read from the remote. And yet the entity's instance which was just a placeholder can be constructed without the presence of the actual remote counterpart.
Now if you use the instance across the code base there is no indication whether the instance of the entity holds the remote counterpart or not. It would make sense if that's optional.
From your implementation it isn't really clear what forms mandatory property and what is optional

Reply:
If mandatory/optional is a concern, I can change `get()` of `RemoteWritableEntity` to return `Optional<T>`

Reply:
Lets simplify this, I don't see a strong reason for this interface implementors to hold any state of the underlying entity, so lets get rid of the get and set methods. Once we do this the implementors are then responsible to only provide ser/de functions.

```
public interface RemoteWriteableEntity<T> {
    /**
     * @return An InputStream created by serializing the entity T
     * @throws IOException Exception encountered while serialization
     */
    public InputStream serialize() throws IOException;

    /**
     * @param inputStream The InputStream which is used to read the serialized entity
     * @return The entity T after deserialization
     * @throws IOException Exception encountered while deserialization
     */
    public T deserialize(InputStream inputStream) throws IOException;

}
```